### PR TITLE
[@mantine/form] useForm: fix isDirty for list fields

### PR DESCRIPTION
**Pull request by gdostie**
By computing the isDirty only on demand using deep comparison of the `values` state, it removes many bugs related to `isDirty` and nested fields. Manually managing a dirty state parallel to the values state was very much bug prone, considering all the possible updates to form values that can occur.

~Disclaimer: by doing this we loose 2 functionalities `initialDirty` and `setDirty`. IMHO the pros outweighs the cons, but I am curious to see what the community thinks.~ Fixed ✅ 

Fixes:  https://github.com/mantinedev/mantine/issues/3001 and https://github.com/mantinedev/mantine/issues/2937

### CODE REVIEW

**Comment by gdostie** [comment]
Converting to draft, will try to retain `initialDirty` and `setDirty`

**Comment by gdostie** [comment]
I managed to retain `setDirty` and `initialDirty` functionalities. 🎉 

I renamed them in the code as "manual overrides" to the standard dirty behaviour.

Reopening for review.

File: src/mantine-form/src/use-form.ts [hunk 1]
```diff
@@ -45,7 +46,7 @@ export function useForm<
   validate: rules,
 }: UseFormInput<Values, TransformValues> = {}): UseFormReturnType<Values, TransformValues> {
   const [touched, setTouched] = useState(initialTouched);
-  const [dirty, setDirty] = useState(initialDirty);
+  const [manualDirtyOverride, setDirtyManualOverride] = useState(initialDirty);
```
Comment (rtivital) [inline src/mantine-form/src/use-form.ts:49]: Is there a reason to rename these variables? I think dirty and setDirty are less noisy

Comment (gdostie) [reply src/mantine-form/src/use-form.ts:49]: The goal was to make it clearer that those are overrides of the normal `initial not equal current` behaviour. I can revert if you feel keeping the original naming this makes sense 👍

**Comment by rtivital** [comment]
Thanks!

**Pull request merged**